src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -111,73 +111,155 @@ public static SocketError Listen(SafeCloseSocket handle, int backlog)
+            }
+        }
+
+        private static unsafe void UnpinBuffers(GCHandle *gcHandles, int count)
+        {
+            for (int i = 0; i < count; i++)
             {
-                objectsToPin = new GCHandle[count];
-                for (int i = 0; i < count; ++i)
+                if (gcHandles[i].IsAllocated)
                 {
-                    ArraySegment<byte> buffer = buffers[i];
-                    RangeValidationHelpers.ValidateSegment(buffer);
-                    objectsToPin[i] = GCHandle.Alloc(buffer.Array, GCHandleType.Pinned);
-                    WSABuffers[i].Length = buffer.Count;
-                    WSABuffers[i].Pointer = Marshal.UnsafeAddrOfPinnedArrayElement(buffer.Array, buffer.Offset);
+                    gcHandles[i].Free();
                 }
+            }
+        }
 
-                // This may throw ObjectDisposedException.
-                SocketError errorCode = Interop.Winsock.WSASend_Blocking(
-                    handle.DangerousGetHandle(),
-                    WSABuffers,
-                    count,
-                    out bytesTransferred,
-                    socketFlags,
-                    SafeNativeOverlapped.Zero,
-                    IntPtr.Zero);
+        private static unsafe byte* GetPinnedBufferPointer(byte[] buffer, int offset)
+        {
+            return (byte*)Marshal.UnsafeAddrOfPinnedArrayElement(buffer, offset);
+        }
+
+        private static unsafe void PopulateWSABufferArray(WSABuffer* wsaBuffers, IList<ArraySegment<byte>> bufferList)
+        {
+            int bufferCount = bufferList.Count;
+            for (int i = 0; i < bufferCount; i++)
+            {
+                ArraySegment<byte> buffer = bufferList[i];
+                byte* bufferPtr = GetPinnedBufferPointer(buffer.Array, buffer.Offset);
+                wsaBuffers[i] = new WSABuffer(bufferPtr, buffer.Count);
+            }
+        }
 
-                if ((SocketError)errorCode == SocketError.SocketError)
+        public static unsafe SocketError Send(SafeCloseSocket handle, IList<ArraySegment<byte>> buffers, SocketFlags socketFlags, out int bytesTransferred)
+        {
+            int bytesSent;
+            int result;
+
+            int count = buffers.Count;
+
+            if (count < WSABuffer.StackAllocLimit)
+            {
+                WSABuffer* wsaBuffers = stackalloc WSABuffer[count];
+                GCHandle* gcHandles = stackalloc GCHandle[count];
+                ClearGCHandleArray(gcHandles, count);
+
+                try
                 {
-                    errorCode = GetLastSocketError();
-                }
+                    PinBuffers(buffers, gcHandles, count);
+                    PopulateWSABufferArray(wsaBuffers, buffers);
 
-                return errorCode;
+                    // This may throw ObjectDisposedException.
+                    result = Interop.Winsock.WSASend_Blocking(
+                        handle.DangerousGetHandle(),
```

Comment:
Seems like we should just be passing the SafeHandle and letting the runtime do the appropriate DangerousAddRef/Release for it, unless we actually need to allow the handle to be closed while this call is in progress to, for example, cause the operation to be canceled?